Fix vim-plugin dependencies

timokau opened the pull request:
###### Motivation for this change

I noticed that plugin dependencies simply did not work with the native vim plugin manager. It simply was not implemented, which was surprising.

To implement it, I first had to refactor vim-utils.nix. Now vim plugins are derivations by default, although the "name + knownPlugins" variant is still possible for backwards compatibility. I can't really see a reason why it was strings in the first place.

I also added a new ncm2 completion backend and updated the plugins (same PR since it would conflict otherwise).

Because of the autogenerated file this will bitrot fast, so I would like to get this merged relatively quickly.

Based on git-shortlog: @andsild, @jagajaga, @Mic92

I have build and tested the provided test instances with the different plugin managers. While doing that I also fixed the pathogen implementation that was already broken. I verified that the installed plugins shows up under `:scriptnames`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

timokau committed:
vimPlugins: Update

timokau committed:
vimPlugins.ncm2-jedi: init at 2018-07-18

timokau committed:
vimUtils: represent vim plugins as derivations

Vim plugins were previously represented as strings by default,
necessitating a `knownPlugins` set. Backwards compatibility is kept
(strings are still accepted), so vam plugins should continue to work.

timokau committed:
vimUtils: add dependency logic to nativeImpl

nativeImpl previously simply ignored dependency information.

timokau commented:
@GrahamcOfBorg build vimUtils.test_vim_with_vim_nix vimUtils.test_vim_with_vim_nix_using_pathogen vimUtils.test_vim_with_vim_nix_using_plug vimPackages.vim-nix vimPackages.ncm2-jedi

timokau commented on pkgs/misc/vim-plugins/overrides.nix:
```diff
@@ -25,14 +25,31 @@ in
+  # This is necessary (as opposed to `with generated;`) because
+  # there would otherwise be conflicts with `rec`. `rec` is necessary
+  # however to make sure overriden plugins can be used as dependencies.
+  # This should use a proper package set with fixed-points instead.
```
@FRidh would it be relatively simple to do that here? Basically `generated.nix` should generate a package set and this should be an overlay to that package set.

Mic92 commented on pkgs/misc/vim-plugins/overrides.nix:
```diff
@@ -177,6 +177,11 @@ with generated;
   });
 
+  ncm2-jedi = ncm2-jedi.overrideAttrs(old: {
+    dependencies = ["nvim-yarp"];
```
Is this not covered by `ncm2` already or is there a way to use ncm2-jedi without?

timokau replied:
It would be covered if `ncm` was a dependency. I wasn't entirely sure if it should be, since I guess the ncm2-jedi provider could in theory be used by another completion plugin.

Also nvim-yarp is used directly, so its nicer to have it as an explicit dependency anyways.

Mic92 replied:
It uses also ncm2's api: https://github.com/ncm2/ncm2-jedi/blob/0418d5ca8d4fe6996500eb04517a946f7de83d34/autoload/ncm2_jedi.vim#L37

timokau replied:
Thats convincing. I added a dependency to ncm2 (still keeping the nvim-yarp dependency because it is also a direct dependency, although it does not matter in practice).

Mic92 commented:
Dependencies is a string for historically reason when the packages where generated by a hacky vim script.

Mic92 commented on pkgs/misc/vim-plugins/vim-utils.nix:
```diff
@@ -183,11 +195,11 @@ let
       (let
         knownPlugins = pathogen.knownPlugins or vimPlugins;
 
-        plugins = map (name: knownPlugins.${name}) (findDependenciesRecursively { inherit knownPlugins; names = pathogen.pluginNames; });
```
For users outside of nixpkgs we should leave at least an assertion, when they still use strings here.

timokau replied:
Strings are explicitly still supported where they were before. As much as I wish we could just remove that, that would break a lot of configs.

Strings are handled in the `transitiveClosure` function which takes an optional `knownPlugins` to resolve them.

timokau replied:
So `findDependenciesRecursively` will always return a list of derivations, even if you input a list of strings.

Mic92 replied:
ok then. I did only read the complete implementation.

timokau commented:
If nobody objects I'll merge this on friday (2018-12-28, Berlin time).

timokau commented:
Just noticed that the other two people I pinged haven't been very active the past year, so probably no point in waiting for a response from them. Are you happy with the current version @Mic92?

Mic92 commented on pkgs/misc/vim-plugins/update.py:
```diff
@@ -310,6 +310,7 @@ def generate_nix(plugins: List[Tuple[str, str, Plugin]]):
                 f"""
   {plugin.normalized_name} = buildVimPluginFrom2Nix {{
     name = "{plugin.normalized_name}-{plugin.version}";
```
Would that work?

```suggestion
    version = plugin.version;
```

Mic92 replied:
Because `stdenv` is computing name anyway.

timokau replied:
Not quite as easily, because `buildVimPlugin` does its own name building. But I have pushed a new commit to add that functionality. I've also refactored the `buildVimPlugin` related functions into its own file, since `vim-utils.nix` is doing too much.

timokau committed:
vimPlugins: Update

timokau committed:
vimUtils.buildVimPlugin: derive name from pname and version

timokau committed:
vimPlugins: use fixed point

timokau commented:
I also figured out how to use fixed points (by copying what python is doing) so its way less hacky now.

timokau commented:
Sorry for all the force-pushes, I don't know how to test eval locally so I'm dependent on ofBorg for that. Should be good to go now.

FRidh commented:
@timokau I use the following `pre-commit` hook:
```sh
#!/bin/sh

echo "Checking evaluation before committing";
exec nix-env -f . -qaP --meta --json --drv-path --show-trace >/dev/null
```

timokau commented:
Thanks, that's very useful.

timokau committed:
vimPlugins.taglist-vim: fix build

The override actually broke the build, no idea why it was added
originally. The build works fine without it.

disassembler commented:
Can we get a release notes update? My neovim plugins are all broken now with the following error, which I'm fine with breaking changes, but lets make sure we have a note in the release notes:

```
building Nix...
building the system configuration...
error: while evaluating the attribute 'activationScript' of the derivation 'nixos-system-sarov-19.03.git.2f218c64ea9' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating the attribute 'system.activationScripts.script' at /home/sam/nixpkgs/custom/nixos/modules/system/activation/activation-script.nix:68:9:
while evaluating 'textClosureMap' at /home/sam/nixpkgs/custom/lib/strings-with-deps.nix:70:35, called from /home/sam/nixpkgs/custom/nixos/modules/system/activation/activation-script.nix:89:18:
while evaluating 'id' at /home/sam/nixpkgs/custom/lib/trivial.nix:14:5, called from undefined position:
while evaluating the attribute 'text' at /home/sam/nixpkgs/custom/nixos/modules/system/activation/activation-script.nix:9:5:
while evaluating the attribute 'text' at /home/sam/nixpkgs/custom/lib/strings-with-deps.nix:77:38:
while evaluating the attribute 'sources' of the derivation 'etc' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating anonymous function at /home/sam/nixpkgs/custom/nixos/modules/system/etc/etc.nix:20:20, called from undefined position:
while evaluating the attribute 'source' at undefined position:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:75:45, called from undefined position:
while evaluating the attribute 'value' at /home/sam/nixpkgs/custom/lib/modules.nix:336:9:
while evaluating the option `environment.etc.systemd/system.source':
while evaluating the attribute 'mergedValue' at /home/sam/nixpkgs/custom/lib/modules.nix:369:5:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:369:32, called from /home/sam/nixpkgs/custom/lib/modules.nix:369:19:
while evaluating 'check' at /home/sam/nixpkgs/custom/lib/types.nix:245:15, called from /home/sam/nixpkgs/custom/lib/modules.nix:370:10:
while evaluating the attribute 'buildCommand' of the derivation 'system-units' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating the attribute 'text' of the derivation 'unit-dbus.service' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating the attribute 'text' at undefined position:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:75:45, called from undefined position:
while evaluating the attribute 'value' at /home/sam/nixpkgs/custom/lib/modules.nix:336:9:
while evaluating the option `systemd.units.dbus.service.text':
while evaluating the attribute 'isDefined' at /home/sam/nixpkgs/custom/lib/modules.nix:374:5:
while evaluating the attribute 'values' at /home/sam/nixpkgs/custom/lib/modules.nix:362:9:
while evaluating the attribute 'values' at /home/sam/nixpkgs/custom/lib/modules.nix:458:7:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:348:28, called from /home/sam/nixpkgs/custom/lib/modules.nix:348:17:
while evaluating 'dischargeProperties' at /home/sam/nixpkgs/custom/lib/modules.nix:416:25, called from /home/sam/nixpkgs/custom/lib/modules.nix:349:62:
while evaluating the attribute 'value' at /home/sam/nixpkgs/custom/lib/modules.nix:232:58:
while evaluating the attribute 'config.text' at /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd.nix:315:7:
while evaluating 'commonUnitText' at /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd.nix:299:20, called from /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd.nix:315:14:
while evaluating 'attrsToSection' at /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd-lib.nix:105:20, called from /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd.nix:301:9:
while evaluating 'mapAttrsToList' at /home/sam/nixpkgs/custom/lib/attrsets.nix:233:23, called from /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd-lib.nix:106:33:
while evaluating the attribute 'unitConfig' at undefined position:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:75:45, called from undefined position:
while evaluating the attribute 'value' at /home/sam/nixpkgs/custom/lib/modules.nix:336:9:
while evaluating the option `systemd.services.dbus.unitConfig':
while evaluating the attribute 'mergedValue' at /home/sam/nixpkgs/custom/lib/modules.nix:369:5:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:369:32, called from /home/sam/nixpkgs/custom/lib/modules.nix:369:19:
while evaluating 'merge' at /home/sam/nixpkgs/custom/lib/types.nix:282:20, called from /home/sam/nixpkgs/custom/lib/modules.nix:372:8:
while evaluating 'filterAttrs' at /home/sam/nixpkgs/custom/lib/attrsets.nix:124:23, called from /home/sam/nixpkgs/custom/lib/types.nix:283:35:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/attrsets.nix:125:29, called from /home/sam/nixpkgs/custom/lib/attrsets.nix:125:18:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/types.nix:283:51, called from /home/sam/nixpkgs/custom/lib/attrsets.nix:125:62:
while evaluating the attribute 'X-Restart-Triggers' at /home/sam/nixpkgs/custom/lib/attrsets.nix:344:7:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/types.nix:283:86, called from /home/sam/nixpkgs/custom/lib/attrsets.nix:344:15:
while evaluating the attribute 'optionalValue' at /home/sam/nixpkgs/custom/lib/modules.nix:376:5:
while evaluating the attribute 'values' at /home/sam/nixpkgs/custom/lib/modules.nix:362:9:
while evaluating the attribute 'values' at /home/sam/nixpkgs/custom/lib/modules.nix:458:7:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:348:28, called from /home/sam/nixpkgs/custom/lib/modules.nix:348:17:
while evaluating 'dischargeProperties' at /home/sam/nixpkgs/custom/lib/modules.nix:416:25, called from /home/sam/nixpkgs/custom/lib/modules.nix:349:62:
while evaluating the attribute 'value' at /home/sam/nixpkgs/custom/lib/types.nix:288:55:
while evaluating the attribute 'X-Restart-Triggers' at /home/sam/nixpkgs/custom/nixos/modules/system/boot/systemd.nix:216:13:
while evaluating the attribute 'serviceDirectories' of the derivation 'dbus-1' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/types.nix:257:14, called from undefined position:
while evaluating the attribute 'value' at /home/sam/nixpkgs/custom/lib/modules.nix:377:27:
while evaluating anonymous function at /home/sam/nixpkgs/custom/lib/modules.nix:369:32, called from /home/sam/nixpkgs/custom/lib/modules.nix:369:19:
while evaluating 'check' at /home/sam/nixpkgs/custom/lib/types.nix:245:15, called from /home/sam/nixpkgs/custom/lib/modules.nix:370:10:
while evaluating the attribute 'passAsFile' of the derivation 'system-path' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating the attribute 'buildCommand' of the derivation 'neovim-0.3.1' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating 'optionalString' at /home/sam/nixpkgs/custom/lib/strings.nix:200:5, called from /home/sam/nixpkgs/custom/pkgs/applications/editors/neovim/wrapper.nix:84:9:
while evaluating the attribute 'passAsFile' of the derivation 'python-2.7.15-env' at /home/sam/nixpkgs/custom/pkgs/stdenv/generic/make-derivation.nix:185:11:
while evaluating 'requiredPythonModules' at /home/sam/nixpkgs/custom/pkgs/top-level/python-packages.nix:94:27, called from /home/sam/nixpkgs/custom/pkgs/development/interpreters/python/wrapper.nix:14:13:
while evaluating 'unique' at /home/sam/nixpkgs/custom/lib/lists.nix:630:12, called from /home/sam/nixpkgs/custom/pkgs/top-level/python-packages.nix:96:6:
while evaluating anonymous function at /home/sam/nixpkgs/custom/pkgs/applications/editors/neovim/wrapper.nix:44:50, called from /home/sam/nixpkgs/custom/pkgs/development/interpreters/python/with-packages.nix:3:19:
while evaluating anonymous function at /home/sam/nixpkgs/custom/pkgs/applications/editors/neovim/wrapper.nix:47:24, called from /home/sam/nixpkgs/custom/pkgs/applications/editors/neovim/wrapper.nix:47:13:
while evaluating anonymous function at /home/sam/nixpkgs/custom/pkgs/applications/editors/neovim/wrapper.nix:41:28, called from undefined position:
while evaluating anonymous function at /home/sam/nixpkgs/custom/pkgs/misc/vim-plugins/vim-utils.nix:427:31, called from undefined position:
value is a set while a string was expected
```

Mic92 commented:
@disassembler breaking the API was not the intention, you should open an issue on that.
